### PR TITLE
Implement User Authentication with JWT for Sign-up, Verification, and Login Endpoints #

### PR DESCRIPTION
**Pull request by jayzalani**
**What does this PR do?**
This PR implements backend functionality for user login with email and password verification. It ensures secure authentication by hashing passwords using the bcrypt library in Python. Please note that verification setup should be customized according to your specific requirements, as I am not aware of the preferred methods you may use. For more information, please refer to the attached demonstration video.

**What issues does this PR fix or reference?**
This PR addresses the issue of securely logging in users with an email and password. It hashes user passwords to enhance security. For further reference on password hashing, you can visit the bcrypt library documentation: https://www.npmjs.com/package/bcrypt

https://github.com/user-attachments/assets/c9ffcfbc-994a-4db6-b846-4eedc4eb6453

### CODE REVIEW

File: store/app/routers/auth/email_auth.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        raise HTTPException(status_code=400, detail="Invalid token")
+    
+    async with UserCrud() as crud:
+        user = await crud.get_user_from_email(email)
+        if user is None:
+            raise HTTPException(status_code=400, detail="User not found")
+        if user.is_verified:
+            return {"message": "Email already verified"}
+        user.is_verified = True
+        await crud._update_item(user)
```
Comment (codekansas) [inline store/app/routers/auth/email_auth.py:64]: shouldn't this be `crud.verify_user`?

File: store/app/routers/auth/email_auth.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+@router.get("/verify")
+async def verify_email(token: str):
+    try:
+        payload = jwt_utils.decode_token(token)
+        email: str = payload.get("sub")
+        if email is None:
+            raise HTTPException(status_code=400, detail="Invalid token")
+    except JWTError:
+        raise HTTPException(status_code=400, detail="Invalid token")
```
Comment (codekansas) [inline store/app/routers/auth/email_auth.py:55]: nit: prefer `status.HTTP_400_BAD_REQUEST`

File: store/app/utils/verify_email.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @abstractmethod
+    async def send_verification_email(self, email: EmailStr, token: str):
+        pass
+
+class ConsoleEmailSender(EmailSender):
+    async def send_verification_email(self, email: EmailStr, token: str):
+        verify_url = f"{settings.site.base_url}/auth/verify?token={token}"
+        logger.info(f"Sending verification email to {email}")
+        logger.info(f"Verification URL: {verify_url}")
+        print(f"\n--- Email ---\nTo: {email}\nSubject: Verify your email\nBody: Please click this link to verify your email: {verify_url}\n--- End Email ---\n")
```
Comment (codekansas) [inline store/app/utils/verify_email.py:18]: assuming this is wip? we do have smtp configured on aws but we haven't been using it

File: store/settings/environment.py [hunk 4]
```diff
@@ -74,3 +76,39 @@ class EnvironmentSettings:
...
+    oauth: OauthSettings = field(default_factory=OauthSettings)
+    user: UserSettings = field(default_factory=UserSettings)
+    crypto: CryptoSettings = field(default_factory=CryptoSettings)
+    email: EmailSettings = field(default_factory=EmailSettings)
+    image: ArtifactSettings = field(default_factory=ArtifactSettings)
+    s3: S3Settings = field(default_factory=S3Settings)
+    dynamo: DynamoSettings = field(default_factory=DynamoSettings)
+    site: SiteSettings = field(default_factory=SiteSettings)
+    auth: AuthSettings = field(default_factory=AuthSettings)  # Add this line
+    debug: bool = field(default=False)
```
Comment (codekansas) [inline store/settings/environment.py:114]: assuming this is dupe?

**Review by codekansas — CHANGES_REQUESTED** [review]
nice pr! few requests, otherwise looks good :)

**Comment by Winston-Hsiao** [comment]
@jayzalani There's been some changes added to master since you started development (see #236). You'll want to merge in current master and resolve the conflicts/re-test some of your signup/login flows. 

There's more covered in the PR #236 comment but main thing is that the User model covers the 3 sign up/login flows. Creation of account (via email/password, Google, and Github) should all create a row in the same User table (normalized model), if using a OAuth provider hashed_password is left empty, and a provider is set (with the OAuth token).

Hopefully you can modify what you've already done to work with the new changes. Let me know if you have any questions also happy to work with you on finishing this up. It looks great so far!

**Comment by chennisden** [comment]
I see that this branch is very WIP.

For now, I have rebased master into this branch. You should see the [jayzalani-rebased](https://github.com/kscalelabs/store/tree/jayzalani-rebased) branch. It has a few fixes as well addressing Ben's comments

**Comment by chennisden** [comment]
A few comments.

You do not need to add `# Add this line` to indicate your diffs. Git does that for you. In particular, it is ideal that you make smaller commits rather than putting all your changes into one big commit.

I have no idea why you merged your branch into itself.

---

Also, because Ben put some other TO-DO stuff in other parts of the code, just a note to everyone: we haven't actuall implemented token verification logic yet. The relevant comment is

`# TODO: Implement token verification logic`

so you can sed for it.

**Comment by jayzalani** [comment]
Understood. I see the issues now. I'll modify the code and retest everything soon. Thanks for the feedback and support

**Comment by chennisden** [comment]
Rebased on master again.

**Pull request closed without merging**